### PR TITLE
feat(cloudfront): apply response-headers policies when serving

### PR DESCRIPTION
**Pull request by abanna**
## Summary

Implements CloudFront response headers policies end to end using AWS REST XML shapes and the policy selected by the matched cache behavior.

- Adds complete create, get, get-config, update, delete, and list behavior for custom and managed response headers policies.
- Parses, validates, stores, and serializes CORS, security, custom, removal, and Server-Timing configuration without flattening nested fields.
- Enforces AWS-compatible ETag handling, exact `Type` values, pagination, name uniqueness, the 20-custom-policy quota, and the 100-distributions-per-policy quota.
- Applies response policies to normal origin responses and custom error responses, with AWS-compatible removal and override ordering.
- Supports managed simple-CORS, preflight-CORS, security, and combined policies.
- Enforces the matched cache behavior's `AllowedMethods`, validates S3 preflight requests, preserves custom-origin behavior, strips hop-by-hop origin headers, and prevents origin `Connection` tokens from suppressing policy-generated headers.
- Implements Server-Timing sampling, `Pragma: server-timing`, and origin metric merging.
- Preserves literal response headers policy IDs in CloudFormation distribution cache behaviors.
- Adds Java SDK and AWS CLI compatibility coverage.

The management and serving changes form one CloudFront response-headers-policy feature; it does not include #1831 or #1832.

## AWS compatibility

Behavior was checked against the official CloudFront and S3 contracts:

- [Response headers policies](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/understanding-response-headers-policies.html)
- [Managed response headers policies](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-response-headers-policies.html)
- [ListResponseHeadersPolicies](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_ListResponseHeadersPolicies.html)
- [CloudFront quotas](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cloudfront-limits.html)
- [AllowedMethods](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_AllowedMethods.html)
- [CloudFront invalid-method behavior](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/http-403-invalid-method.html)
- [S3 OPTIONS requests](https://docs.aws.amazon.com/AmazonS3/latest/developerguide/RESTOPTIONSobject.html)

The custom `AWS::CloudFront::ResponseHeadersPolicy` CloudFormation resource is not implemented in this change. Distribution templates can reference literal managed or custom policy IDs.

## Verification

- Full repository suite: 8,218 tests, 0 failures, 0 errors, 9 skipped.
- Complete CloudFront and response-headers-policy test family: 118 tests, 0 failures, 0 errors.
- Java AWS SDK lifecycle round trip: 1 test passed across create, get, get-config, list, stale-ETag update, update, delete, and modeled errors.
- Real AWS CLI list validation passed for modeled payload decoding and multi-page `NextMarker` traversal.
- Strict action-table documentation generation passed.
- Documentation tooling tests: 20 passed.
- Exact staged artifact audit passed for whitespace, conflicts, scope, generated files, and publication boundaries.

## Type

- [x] New feature

### CODE REVIEW

**Comment by abanna** [comment]
The current native / sdk-test-java failure is outside this PR's CloudFront response-headers diff. It is CodeBuildTest.retryBuild expecting the retried build response to remain IN_PROGRESS but receiving FAILED. The focused compatibility fix is #2052; after that lands, this job should be rerun or the branch should be refreshed.

**Comment by abanna** [comment]
### Reviewing this PR

The diff on this page is **#1820 + this PR**, because this branch contains #1820's base commit `784359b`. The change to review here is the single commit [`7a70680`](https://github.com/floci-io/floci/pull/1833/commits/7a70680485d507eeb8303f72225c4127fddb58f1) — **22 files, 3,077 insertions**.

It cannot be retargeted to stack on #1820: this is a cross-repo PR from a fork, and GitHub only accepts a base branch that exists in `floci-io/floci`. Once #1820 merges this page collapses to its own commit on its own.

#1831, #1832 and #1833 are siblings on #1820, not a chain, and all three conflict pairwise — so they need to merge one at a time with the others re-merged in between. The fan diagram, the conflict table and a suggested order are in [this comment on #1820](https://github.com/floci-io/floci/pull/1820#issuecomment-5200048904).

File: src/main/java/io/github/hectorvent/floci/services/cloudfront/CloudFrontService.java [hunk 1]
```diff
@@ -546,6 +599,178 @@ public List<ResponseHeadersPolicy> listResponseHeadersPolicies(String marker, in
...
+        return defaultUsesPolicy || orderedUsesPolicy;
+    }
+
+    private static Map<String, ResponseHeadersPolicy> managedResponseHeadersPolicies() {
+        Map<String, Object> simpleCors = corsConfig(
+                List.of("*"), List.of(), List.of(), List.of(), null);
+        Map<String, Object> preflightCors = corsConfig(
+                List.of("*"),
+                List.of("GET", "HEAD", "PUT", "POST", "PATCH", "DELETE", "OPTIONS"),
+                List.of(), List.of("*"), null);
```
[the remark this answers is not in the document]
Comment (abanna) [reply src/main/java/io/github/hectorvent/floci/services/cloudfront/CloudFrontService.java:711]: Thanks for the review. I checked this against the [official AWS managed response headers policy definition](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-response-headers-policies.html) and verified the live AWS-managed `CORS-With-Preflight` policy through `GetResponseHeadersPolicy`; `AccessControlAllowHeaders` returns `Quantity: 0`. AWS defines allow-methods, allow-origin `*`, and expose-headers `*` for this policy, but not allow-headers. The current implementation and regression test intentionally match that AWS-managed shape. Adding `Access-Control-Allow-Headers: *` would create a protocol divergence, so no code change is needed. Thank you!

**Comment by pgermosen** [comment]
Hi @abanna, thorough piece of work here — reviewed the current state directly rather than relying on the earlier review thread.

**Merge conflict.** Confirmed with an actual merge attempt (not just the GitHub status indicator): one real conflict, in `Ec2PublicIpOnLaunchTest.java`. It's small and unrelated to CloudFront — looks like a stray pickup from a branch sync rather than anything intentional. Could you rebase onto current `main`?

**The four code-quality findings from the review thread are genuinely fixed**, verified by reading the current code rather than trusting the thread:
- The swallowed XML parse exception in `ResponseHeadersPolicyConfigCodec` is actually handled better than the original suggestion — malformed XML is now logged *and* rejected outright with `InvalidArgument` (400) via the new shared `XmlParser.extractElementTree`, rather than silently producing an empty policy.
- `Access-Control-Allow-Origin` no longer goes through the comma-joining `corsListHeader` path — it's now built from `selectAllowOrigin` returning a single matching origin, which is spec-correct (`AccessControlExposeHeaders`/`AccessControlAllowHeaders` correctly still use the joined form, since those *do* support comma-separated values per the Fetch spec).
- `ResponseHeadersPolicyId` is now read in both `cfnDefaultCacheBehavior` and `cfnCacheBehaviors`.
- `parseCustomErrorResponses`'s catch block now logs via `LOG.debugv` instead of swallowing silently.
- The `Access-Control-Allow-Headers`-missing-from-managed-preflight-policy thread resolved itself correctly — your live `GetResponseHeadersPolicy` verification (`Quantity: 0`) settled it, and Greptile conceded.

**The one still-open item is the P0 SSRF finding** on `fetchFromCustomOrigin` — confirmed real: `origin.getDomainName()` flows straight into the outbound request URI with no allow-list, private-IP/metadata-endpoint check, or DNS-rebinding guard. That said, I checked and this isn't unique to this PR — `HttpProxyInvoker` (API Gateway's `HTTP_PROXY` integration) builds its outbound target from `IntegrationUri` the exact same way, with the same lack of validation. So this looks like an existing, pervasive pattern across every "forward to a caller-specified origin" feature in the codebase, not a regression this PR introduces. I don't think it's fair to block this PR alone on fixing it, but it's worth a maintainer call on whether it needs addressing here, as a separate follow-up, or is an accepted tradeoff for a local emulator's threat model.

Ran the full CloudFront + response-headers-policy + XmlParser test families locally: 154 tests, 0 failures. `docs-check` passes, no CODEOWNERS-protected paths touched.

So: ready modulo the rebase, and worth a decision on the SSRF item before merging.

**Review by pgermosen — APPROVED** [review]
@abanna Approving. 223/223 green locally on `356d7997`, and 223/223 on the tree merged with current `main` — verified with a real merge, not the cached status. The rebase I asked for landed cleanly: `Ec2PublicIpOnLaunchTest.java` is byte-identical to `main` and out of the diff.

## Correcting my earlier review

**The SSRF finding I raised on 2026-08-18 was wrong, and it was the thing I said should block.** I claimed `origin.getDomainName()` flows into the outbound request "with no allow-list, private-IP/metadata-endpoint check, or DNS-rebinding guard." All three exist, and they were already on `main` when I wrote that — #1820 added them on 2026-08-12, six days earlier. I should have found them before asking you to wait on a maintainer decision.

`CloudFrontOriginHttpClient` installs a `ValidatingDnsResolver` as the connection manager's DNS resolver, and it screens the **resolved addresses** through `isBlockedOriginAddress`: loopback, any-local, link-local (so `169.254.169.254` is covered), site-local, multicast, CGNAT `100.64/10`, reserved `240/4`, IPv6 unique-local `fc00::/7`, documentation `2001:db8::/32`, and the IPv4-mapped form of each.

Screening at the resolver is the strong placement rather than an incidental one: it checks the addresses actually used to open the connection, so DNS rebinding is defeated by construction instead of by a pre-flight lookup that a second resolution could invalidate. Redirect following is off (`disableRedirectHandling` plus `setRedirectsEnabled(false)`), which closes the redirect-to-internal bypass. And the escape hatch is `allowed-private-origin-hosts`, an operator config key that ships empty — not something a distribution's own `DomainName` can populate.

So there was no decision to make. This PR doesn't touch that method either; I diffed the body against `main` and it is identical.

## What I verified

Against the CloudFront docs and the botocore `2020-05-31` model, not your description:

- **All five managed policy IDs are exact** — `60669652-…` SimpleCORS, `5cc3b908-…` CORS-With-Preflight, `67f7725c-…` SecurityHeadersPolicy, `e61eb60c-…` CORS-and-Security, `eaab4381-…` CORS-with-preflight-and-Security.
- **Every error code is a real modeled shape at the right status, thrown only on the operation that declares it**: `TooManyResponseHeadersPolicies` 400, `ResponseHeadersPolicyAlreadyExists` 409, `ResponseHeadersPolicyInUse` 409, `IllegalUpdate` 400 (Update only), `IllegalDelete` 400 (Delete only), `PreconditionFailed` 412, `NoSuchResponseHeadersPolicy` 404, and `TooManyDistributionsAssociatedToResponseHeadersPolicy` 400 on both Create and Update.
- **The ETag split is right.** `InvalidIfMatchVersion` 400 for a missing or empty `If-Match` in the controller, `PreconditionFailed` 412 for a present-but-stale one in the service. AWS lists both on Update and Delete, and that is precisely the distinction between them — easy to collapse into one code, and you didn't.
- **The remove-headers denylist matches the doc entry for entry**: 31 literals plus the `x-amz-cf-` and `x-edge-` prefixes, 33 total.
- **Quotas match**: 20 custom policies, 100 distributions per policy, 10 custom headers, 1,783-char CSP, sampling 0–100 with four decimals.
- **Remove-then-add ordering**, **`Pragma: server-timing` forcing the header at sampling rate 0**, and **merging rather than replacing the origin's `Server-Timing`** all match the docs.
- **`Server: CloudFront` substitution after removing `Server`/`Date`** is a precise catch — that behavior lives in a footnote on the remove-headers page and is the kind of thing that normally gets missed.
- **The `AccessControlAllowOrigins` wildcard rule** implements leftmost-subdomain-only correctly: `*.example.org` accepted; `example.*`, `test.*.example.org`, `*test.example.org` and `exa*mple.org` rejected.

Ten mutations, nine caught by a named test — per-header `Override`, the `Server`/`Date` substitution, `AllowedMethods` 403, the `Server-Timing` merge, the `Pragma` override, managed-policy immutability, the in-use delete guard, the 20-policy quota, and the 100-distributions-per-policy quota. Worth singling out: regressing `selectAllowOrigin` back to the comma-joined form now fails `allowOriginIsASingleValueNotACommaJoinedList` — a regression test named after the bug it prevents. That's the habit that makes a fix stay fixed.

## Follow-up: the CORS group-override rule has no test

One mutation survived. I deleted the group-override suppression outright — `isCorsHeader(normalized) && !header.override() && hasOriginCorsHeader` — and all 185 tests still passed.

The rule is *correct*. The doc is explicit:

> "When set to `false` and the origin response contains a CORS header (regardless of whether the CORS header is in the policy), CloudFront includes the CORS header that it received from the origin to the response. CloudFront doesn't add any CORS headers in the policy to the response that is sent to the viewer."

That is group-wide, and your implementation is group-wide. Nothing to change in the behavior.

The gap is that nothing distinguishes it from ordinary per-header semantics. `responseHeadersPolicyRemovesOverridesAndPreservesOriginHeaderValues:169` looks like coverage — the origin sends `Access-Control-Allow-Origin` and the test asserts it survives — but that assertion passes under both rules, since the per-header branch would preserve it too.

The distinguishing case is: **the origin sends CORS header A, and the policy supplies a different CORS header B with `OriginOverride=false`.** AWS adds neither. Per-header semantics would add B. Concretely: origin returns `Access-Control-Allow-Origin`, policy's `CorsConfig` sets `AccessControlMaxAgeSec`, and the response must carry no `Access-Control-Max-Age`.

That's the most subtle piece of AWS semantics in a 3,000-line change, and right now it's the one piece that can regress without a test noticing. A single test closes it.

## Non-blocking

1. **`PolicyHeader` can't tell a CORS header from a custom one.** The record is `(name, value, override)` with no source tag, so a `CustomHeadersConfig` entry named `access-control-*` gets swept into the CORS group rule, where the doc gives custom headers per-header semantics. Greptile flagged this and rated it unlikely — it is — but adding a `cors` flag to `PolicyHeader` and gating the group rule on it also gives you the natural hook for the test above, so the two are worth doing together.

2. **`AccessControlAllowHeaders` wildcard with `AccessControlAllowCredentials=true` is rejected**, pinned by `rejectsWildcardAllowHeadersWhenCredentialsAreAllowed`. That's a correct *Fetch spec* rule, but the CloudFront docs don't say `CreateResponseHeadersPolicy` rejects it — the only wildcard caveat there is that `Authorization` can't use one and must be listed explicitly. This is the same shape as the `authorization` denylist entry on #1832: an undocumented rejection of a config real AWS may accept, which is the worse direction for an emulator to diverge. If you probed real CloudFront for it, cite that and it's settled; otherwise it's worth dropping.

3. **Two limits are inferred rather than doc-backed** — `MAX_REMOVE_HEADERS = 10` and the 11-entry `FORBIDDEN_CUSTOM_HEADERS` set. Both are reasonable (symmetry with the custom-header quota, and the hop-by-hop set), just not in the quota table like everything else here.

4. **The `KinesisIntegrationTest` change is unrelated scope but correct.** `BigDecimal.equals` compares scale as well as value, so a millisecond timestamp ending in `0` makes the string parse and Jackson's `decimalValue()` agree on value and disagree on scale — a real intermittent failure, roughly one run in ten. `compareTo` is the right comparison. Worth a line in the description saying why it's in this PR.

## Merge ordering

#1832 and #1833 touch nine of the same files and **conflict pairwise** — I confirmed by merging them, with conflicts in `CloudFrontService`, `CloudFrontServingController`, `CloudFrontRequestRouter`, `CloudFormationCloudFrontDistributionIntegrationTest`, `CloudFrontDistributionServingTest` and `CloudFrontServiceTest`. Both report `CLEAN` only because each is clean against `main` on its own. Whichever lands first, the other needs a re-merge before it can go in.

Approving now; the CORS group-override test can come as a follow-up.

**Comment by pgermosen** [comment]
🎉 Merged — thanks for the thorough work on this, @abanna, including chasing down the CORS group-override edge case with a precise test. This was a big, careful piece of AWS-fidelity work.
